webstruct/text_tokenizers.py

```diff
@@ -79,46 +143,66 @@ class WordTokenizer(object):
     def _tokenize(self, text):
         # this one cannot be placed in the loop because it requires
         # position check (beginning of the string) or previous char value
-        text = self.open_quotes.sub(r'\1``', text)
+        quote = self.open_quotes.search(text)
+        if quote is not None:
+            end = quote.end() - 1
+            for t in self._tokenize(text[:end]):
+                yield t
+            yield TextToken(chars='``', position=end, length=1)
+            shift = end + 1
+            for t in self._tokenize(text[shift:]):
+                yield TextToken(chars=t.chars,
+                                position=t.position + shift,
+                                length=t.length)
+            return
 
         i = 0
         token_start = 0
         while 1:
             if i >= len(text):
-                yield text[token_start:]
+                yield TextToken(chars=text[token_start:],
+                                position=token_start,
+                                length=len(text) - token_start)
                 break
             shift = 1
             partial_text = text[i:]
             for regex, token in self.rules:
                 match = regex.match(partial_text)
                 if match:
-                    yield text[token_start:i]
+                    yield TextToken(chars=text[token_start:i],
+                                    position=token_start,
+                                    length=i - token_start)
                     shift = match.end() - match.start()
                     token_start = i + shift
                     if token is None:
-                        yield match.group()
+                        yield TextToken(chars=match.group(),
+                                        position=i + match.start(),
+                                        length=shift)
                     else:
-                        yield token
+                        yield TextToken(chars=token,
+                                        position=i + match.start(),
+                                        length=shift)
                     break
             i += shift
 
     def tokenize(self, text):
-        return [t for t in self._tokenize(text) if t]
+        return [t for t in self._tokenize(text) if t.chars]
```

Comment:
Does it make https://github.com/scrapinghub/webstruct/pull/36/files obsolete?

Also, it seems we can easily make this backwards compatible by using another name for tokenize, and leaving existing tokenize method as-is.

Reply:
Yes, it does same what span want to do. But 36th pull request also contains some regexp modifications.

You want to move new code to separate method, e.g. span_tokenize, and keep old method tokenize calling for span_tokenize internally, which will be return only text tokens? This is possible, but there is no consumers for old method.

Reply:
I think we can keep regexp modifications out of this PR; #36 was not merged because quality was decreasing a bit after tokenization changes.

As for consumers, that's true there are no consumers in webstruct, but we never know as it is open source :) Also, tokenizer is made to be compatible with nltk's tokenizer; it is not a good design to return a different data type in an overridden method.